### PR TITLE
snapshots: import of a snapshot set

### PR DESCRIPTION
**Pull request by slimjim777**
Partner feature for PR #8982 

Imports an exported snapshot set to a different device. The snapshot fileset is unpacked to disk and then moved into place with a local snapshot ID. The snapshots are not restored - that needs to be done by the user via `snap restore`.

Note: added for early review, but needs content file size checks and spread tests

### CODE REVIEW

File: overlord/snapshotstate/backend/backend.go [hunk 1]
```diff
@@ -323,3 +346,116 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+			return nil, err
+		}
+		snaps = append(snaps, snapshot.Snap)
+
+		// set the new setID and get the new filename
+		snapshot.SetID = id
+		new := Filename(snapshot)
+		old := path.Join(p, name)
+
+		if err := os.Rename(old, new); err != nil {
```
Comment (bboozzoo) [inline overlord/snapshotstate/backend/backend.go:456]: Don't we need to update the set ID in the metadata as well? AFAICT right now it will carry the same ID that was set during export.

File: vendor/vendor.json [hunk 2]
```diff
@@ -110,7 +110,7 @@
 			"revisionTime": "2017-09-28T14:21:59Z"
 		},
 		{
-			"checksumSHA1": "fqejS2llZXw3gLnOYhg7pcSlY+Q=",
+			"checksumSHA1": "Yfj2ZrgRrklXrshCM1RxH5pvUY8=",
```
Comment (bboozzoo) [inline vendor/vendor.json:113]: This change should not be part of the PR.

File: overlord/snapshotstate/backend/backend.go [hunk 3]
```diff
@@ -157,6 +160,26 @@ func Filename(snapshot *client.Snapshot) string {
...
+	var setID uint64
+	var revision int
+	var snapName, version string
+	parseF := strings.Replace(f, "_", " ", 3)
+	parseF = strings.Replace(parseF, ".zip", "", 1)
+	if _, err := fmt.Sscanf(parseF, "%d %s %s %d", &setID, &snapName, &version, &revision); err != nil {
+		return nil, fmt.Errorf("unexpected filename format: %v", err)
+	}
+
+	snapshot := &client.Snapshot{
```
[the remark this answers is not in the document]
Comment (slimjim777) [reply overlord/snapshotstate/backend/backend.go:173]: Sounds good. Would that work if a snapshot is being restored on a different machine (or one that has been factory reset)?

File: overlord/snapshotstate/backend/backend.go [hunk 4]
```diff
@@ -323,3 +346,116 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 
 	return nil
 }
+
+// Import a snapshot from the export file format
+func Import(ctx context.Context, id uint64, r io.Reader) (int64, []string, error) {
+	comment := fmt.Sprintf("snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	p := path.Join(dirs.SnapCacheDir, "snapshots", string(id))
```
[the remark this answers is not in the document]
Comment (slimjim777) [reply overlord/snapshotstate/backend/backend.go:355]: I've used the cache directory, but I can change the name to `import-<id>`. Did you mean that this should be created in `dirs.SnapshotsDir`?

File: overlord/snapshotstate/backend/backend.go (hunk 1 above)
Comment (slimjim777) [reply overlord/snapshotstate/backend/backend.go:456]: That is set in the previous line:
```
		// set the new setID and get the new filename
		snapshot.SetID = id
		new := Filename(snapshot)
		old := path.Join(p, name)
```

File: vendor/vendor.json (hunk 2 above)
Comment (slimjim777) [reply vendor/vendor.json:113]: hmmm... not sure how that changed

File: overlord/snapshotstate/backend/backend.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapshotstate/backend/backend.go:355]: Ah, right, sorry I missed that. Please ignore that comment.

File: overlord/snapshotstate/backend/backend.go (hunk 1 above)
Comment (bboozzoo) [reply overlord/snapshotstate/backend/backend.go:456]: I meant the ID stored in the metadata file inside the snapshot itself. We import the snapshot and save if under a name that reflects the ID one request but AFAICT the metadata is not updated. From looking at other code, I think it will also break how snapshots are lited in `snap saved`.

File: overlord/snapshotstate/backend/backend.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapshotstate/backend/backend.go:173]: It should, from what I remember snapshots should be self contained.

File: overlord/snapshotstate/backend/backend.go (hunk 1 above)
Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go:504]: Yes, @bboozzoo is right. Unfortunately only the set id stored in the metadata inside the zip matters atm (updating it would require updating of the zip file). I'm working on changing that.

File: overlord/snapshotstate/backend/backend.go [hunk 5]
```diff
@@ -178,6 +178,27 @@ func Filename(snapshot *client.Snapshot) string {
 	return filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_%s_%s_%s.zip", snapshot.SetID, snapshot.Snap, snapshot.Version, snapshot.Revision))
 }
 
+func setIDFromFilename(f string) (uint64, error) {
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:181]: I think it's unfortunate if the filename and the json disagree about the snapshot-id, We need to do something about this and this helper needs to go. I think @stolowski has some plans here fortunately :)

Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:181]: Actually talking to @stolowski the filename will be the authority here and the json will contain "0" so this is good.

File: overlord/snapshotstate/backend/backend.go [hunk 6]
```diff
@@ -412,6 +424,118 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+	// unpack the tar file to a temporary location (so the contents can be validated)
+	exportFound, size, err := unpackSnapshotImport(r, p)
+	if err != nil {
+		return 0, nil, err
+	}
+	if !exportFound {
+		return 0, nil, fmt.Errorf("snapshot import file incomplete: no export.json file")
+	}
+
+	// XXX: check the snapshot hashes (these are not in export.json at present)
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:453]: I think we have hashes now fwiw

File: overlord/snapshotstate/backend/backend.go [hunk 7]
```diff
@@ -412,6 +424,118 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+		}
+
+		target := path.Join(p, header.Name)
+		t, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+		if err != nil {
+			return false, 0, fmt.Errorf("failed creating file `%s`: %v", target, err)
+		}
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:505]: I think we leak the "t" file descriptor here if we exit here.

File: overlord/snapshotstate/backend/backend.go (hunk 1 above)
Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:532]: This is now changed and the sid from the filename is now the authoritative part and the metadata is ignored.

**Review by mvo5 — COMMENTED** [review]
Added some more comments that need attention but I think this is getting closer and closer :)

File: overlord/snapshotstate/backend/backend.go [hunk 8]
```diff
@@ -445,6 +457,118 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+
+	// XXX: check the snapshot hashes (these are not in export.json at present)
+
+	// walk the cache directory to store the files
+	dir, err := osOpen(p)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed opening import cache: %v", comment)
+	}
+	defer dir.Close()
+	names, err := dirNames(dir, 100)
```
Comment (stolowski) [inline overlord/snapshotstate/backend/backend.go:494]: This returns up to 100 names, but needs to be repeated until EOF.

File: overlord/snapshotstate/backend/backend.go (hunk 7 above)
Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go:539]: Fixed.

File: overlord/snapshotstate/backend/backend.go (hunk 8 above)
Comment (pedronis) [reply overlord/snapshotstate/backend/backend.go:482]: yes

File: overlord/snapshotstate/backend/backend.go [hunk 9]
```diff
@@ -445,6 +457,117 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+		old := path.Join(p, name)
+		snapshot, err := snapshotFromFilename(old)
+		if err != nil {
+			return nil, err
+		}
+		snaps = append(snaps, snapshot.Snap)
+
+		// set the new setID and get the new filename
+		snapshot.SetID = id
+		new := Filename(snapshot)
```
Comment (pedronis) [inline overlord/snapshotstate/backend/backend.go:553]: wondering if a FilenameWithSetID API would be a bit nicer

File: overlord/snapshotstate/backend/backend.go [hunk 10]
```diff
@@ -445,6 +445,119 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
+		}
+		size += writtenSize
+	}
+	return exportFound, size, nil
+}
+
+func moveCachedSnapshots(names []string, id uint64, p string) ([]string, error) {
```
Comment (pedronis) [inline overlord/snapshotstate/backend/backend.go:534]: this should identify if we have a snapshots already with the same data using content hashes defined leaving out set id and timestamps

**Review by pedronis — COMMENTED** [review]
some comments

File: overlord/snapshotstate/backend/backend.go [hunk 11]
```diff
@@ -476,42 +476,53 @@ func Import(ctx context.Context, id uint64, r io.Reader) (int64, []string, error
...
-	}
 
-	// move the files into place with the new local set ID
-	snapNames, err := moveCachedSnapshots(names, id, p)
+	var snapNames []string
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if err != nil {
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:486]: Hm, what `err` is this checking? readErr? or a leftover err?

Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go:486]: Thanks for catching, fixed.

File: overlord/snapshotstate/backend/backend.go [hunk 12]
```diff
@@ -445,6 +446,129 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
 }
 
+// Import a snapshot from the export file format
+func Import(ctx context.Context, id uint64, r io.Reader) (size int64, snapNames []string, err error) {
+	comment := fmt.Sprintf("snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	p := path.Join(dirs.SnapCacheDir, "snapshots", fmt.Sprintf("import-%d", id))
+	if err := os.MkdirAll(path.Join(dirs.SnapCacheDir, "snapshots"), 0755); err != nil {
+		return 0, nil, fmt.Errorf("cannot create snapshots directory")
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go]: This should return the reason for the error

File: overlord/snapshotstate/backend/backend.go [hunk 13]
```diff
@@ -445,6 +446,129 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+	}
+	defer dir.Close()
+
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if len(names) > 0 {
+			// move the files into place with the new local set ID
+			names, err := moveCachedSnapshots(names, id, p)
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go]: This needs to undo the partial imported files on failure. i.e. if moveCachedSnapshot files fails after the 2nd imported snapshot file right now we don't cleanup the first imported file AFAICT

File: overlord/snapshotstate/backend/backend.go [hunk 14]
```diff
@@ -445,6 +446,129 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
+		}
+		size += writtenSize
+	}
+	return exportFound, size, nil
+}
+
+func moveCachedSnapshots(names []string, id uint64, p string) ([]string, error) {
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:544]: This should actually just move things around from one dir to another, currently it's doing much more (validating, renaming).

File: overlord/snapshotstate/backend/backend.go [hunk 15]
```diff
@@ -445,6 +446,129 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+	if err := os.Mkdir(p, 0755); err != nil {
+		if os.IsExist(err) {
+			return 0, nil, fmt.Errorf("snapshot import already in progress for ID `%d`", id)
+		}
+		return 0, nil, fmt.Errorf("failed creating import cache: %v", err)
+	}
+	defer os.RemoveAll(p)
+
+	// unpack the tar file to a temporary location (so the contents can be validated)
+	exportFound, size, err := unpackSnapshotImport(r, p)
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go]: Either here or in a new helper we need to do the validate the import (Check, hashes, rename)

**Review by mvo5 — COMMENTED** [review]
Another pass, still needs some love :) will see what I can do after lunch

File: overlord/snapshotstate/backend/backend_test.go [hunk 16]
```diff
@@ -887,22 +890,23 @@ func (s *snapshotSuite) TestMaybeRunuserNoHappy(c *check.C) {
 	c.Check(strings.TrimSpace(logbuf.String()), check.Matches, ".* No user wrapper found.*")
 }
 
+func parseSnapshotFilename(c *check.C, fname string) (snapName, version string, revision int) {
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend_test.go:897]: :heart:

File: overlord/snapshotstate/backend/backend.go [hunk 17]
```diff
@@ -445,6 +446,149 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+
+		old := path.Join(cachedSnapshotsCDir, name)
+
+		// read old snapshot, override its set id internally
+		r, err := backendOpen(old, id)
+		if err != nil {
+			return nil, fmt.Errorf("cannot open snapshot: %v", err)
+		}
+		r.Close()
+		if err != nil {
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:576]: We don't need this `if err != nil` now anymore AFAICT after the latest refactor that moves check out of this helper.

File: cmd/snap/cmd_help.go [hunk 18]
```diff
@@ -220,7 +220,7 @@ var helpCategories = []helpCategory{
 	}, {
 		Label:       i18n.G("Snapshots"),
 		Description: i18n.G("archives of snap data"),
-		Commands:    []string{"saved", "save", "check-snapshot", "restore", "forget"},
+		Commands:    []string{"saved", "save", "check-snapshot", "restore", "forget", "export-snapshot", "import-snapshot"},
```
Comment (pedronis) [inline cmd/snap/cmd_help.go:223]: these should maybe go to AllOnlyCommands for now

File: overlord/snapshotstate/backend/backend.go [hunk 19]
```diff
@@ -445,6 +446,175 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+		}
+
+		target := path.Join(p, header.Name)
+		if header.Name == "export.json" {
+			exportFound = true
+		} else {
+			extractedSnapshots = append(extractedSnapshots, target)
+		}
+
+		t, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
```
Comment (pedronis) [inline overlord/snapshotstate/backend/backend.go:547]: we should pick our mode and not use the header I think

File: overlord/snapshotstate/backend/backend_test.go [hunk 20]
```diff
@@ -1043,3 +1233,111 @@ func (s *snapshotSuite) TestExportUnhappy(c *check.C) {
...
+		}
+		if _, err := tw.Write(buf.Bytes()); err != nil {
+			return err
+		}
+	}
+
+	if flags.withDir {
+		hdr := &tar.Header{
+			Name:     dirs.SnapshotsDir,
+			Mode:     0755,
```
Comment (pedronis) [inline overlord/snapshotstate/backend/backend_test.go:1315]: this looks wrong compared to reality

**Review by pedronis — COMMENTED** [review]
did another pass, some further comments

File: overlord/snapshotstate/backend/backend.go (hunk 17 above)
Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go]: I fixed this now

File: overlord/snapshotstate/backend/backend.go [hunk 21]
```diff
@@ -445,6 +446,169 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
...
+	}
+	defer dir.Close()
+
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if len(names) > 0 {
+			// move the files into place with the new local set ID
+			newSnapNames, newFiles, err := moveCachedSnapshots(tempImportDir, names, id)
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:501]: This is not an atomic operation. If we reboot in the middle of this move we will have half an imported snapshot but no indication that the move was incomplete. Also right now on core this move from /var/cache/ to /var/lib gives us an "invalid cross-device link" error which looks like a fun side-effect of our writable-path magic.

Comment (pedronis) [reply overlord/snapshotstate/backend/backend.go:501]: seems we need to discuss what to do about these points

Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:501]: I chatted with Samuele and we will fix this by having a "$sid_importing" file that will be created first. The backend list code will skip all files with an importing setID. When the improt is finished this file is removed again.

**Comment by mvo5** [comment]
Fwiw, I also pushed #946, #9462 and #9467 with some small tweaks that will be useful here.

File: overlord/snapshotstate/backend/backend.go (hunk 21 above)
Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:566]: The commit https://github.com/snapcore/snapd/pull/9036/commits/7f833c2c5b5d751be2b67204e4166e4e891553e7 has a sketch of my proposal to solve the issue, namely:
1. the Iter() code ignores any setID that has a $setID_importing filename
2. when we start importing we write *first* a $setID_importing filename
3. we import into the regular snapshot dir, this avoid all cross-device moves or non-atomic multiple-files moves
4. we either Commit() or Cancel(), i.e. either remove the $setID_importing filename (commit) or do `rm $setID_*.zip` to cancel.
5. on startup we find all leftover $setID_imprting files and cleanup the "$setID_*.zip"

Comment (zyga) [reply overlord/snapshotstate/backend/backend.go:566]: @mvo5 consider just using file locks, they seem like an easier solution as this removes 5. (on startup they are not locked), extra stat calls in 2, etc.

Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:566]: @zyga happy to chat about this but (5) we will always need, what if we restart in the middle of an import? we always need to cleanup

Fwiw, I pushed a full version (minus tests and cleanup on startup) in https://github.com/mvo5/snappy/tree/snapshot-import-3

Comment (zyga) [reply overlord/snapshotstate/backend/backend.go:566]: I misread point 5., please disregard my comment on that.

**Comment by mvo5** [comment]
Closing in favor of #9469  which wil be merged once spread has finished :)
